api/common/io_utils.go

```diff
@@ -34,6 +34,39 @@ func (g *clampWriter) Write(p []byte) (int, error) {
 	return n, err
 }
 
+type clampReadCloser struct {
+	r           io.ReadCloser
+	remaining   int64
+	overflowErr error
+}
+
+func NewClampReadCloser(buf io.ReadCloser, max uint64, overflowErr error) io.ReadCloser {
+	if max != 0 {
+		return &clampReadCloser{r: buf, remaining: int64(max), overflowErr: overflowErr}
+	}
+	return buf
+}
+
+func (g *clampReadCloser) Close() error {
+	return g.r.Close()
+}
+
+func (g *clampReadCloser) Read(p []byte) (int, error) {
+	if g.remaining <= 0 {
+		return 0, g.overflowErr
+	}
+	if int64(len(p)) > g.remaining {
+		p = p[0:g.remaining]
+	}
+
+	n, err := g.r.Read(p)
+	g.remaining -= int64(n)
+	if g.remaining <= 0 {
```

Comment:
Not sure if your callers expect this, but if `len(p)` is 1024, and `max` was 1024, then a Read(p) will return `1024, g.overflowErr` on the first invocation.  If the callers are checking for err being non-nil, this may confuse them.  OTOH, if they're checking size before checking for errors then this is probably fine.

Reply:
Personally, I find confusing returning the error if the case that @kmjohansen described. Looking at the implementation of MaxBytesReader in the case (max=len(p)) it doesn't return an error: https://golang.org/src/net/http/request.go?s=32861:32938#L1038 (unless the Read itself returns an error). So I think people don't expect to get an error back in that case, what do you think? Once we take a decision it would be great to have a simple unit test to assert what we have decided, implementing the different scenarios.

Reply:
I agree, reading up to and including the full max size should be fine.